### PR TITLE
NOZ-49: Decode AND

### PR DESCRIPTION
**Pull request by nozai-adam**
## Summary

Adds support for decoding RISC-V AND instructions as part of expanding the instruction decoder capabilities.

## Changes

- **Instruction Support**: Added AND instruction variant to the RiscVInstruction enum with proper decode logic (funct3=0x7, funct7=0x00)
- **Display Formatting**: Implemented display formatting for AND instruction output
- **Test Coverage**: Added comprehensive unit tests for both instruction decoding and display formatting
- **Code Organization**: Maintained existing test structure with separate decode and display test modules

This change enables the RISC-V decoder to handle bitwise AND operations, maintaining full test coverage (290/290 tests passing) and following the established code organization patterns.

### CODE REVIEW

File: riscv/src/instruction.rs [hunk 1]
```diff
@@ -225,6 +234,8 @@ const XOR_FUNCT3: u8 = 0x4;
 const XOR_FUNCT7: u32 = 0x00;
 const OR_FUNCT3: u8 = 0x6;
 const OR_FUNCT7: u32 = 0x00;
+const AND_FUNCT3: u8 = 0x7;
+const AND_FUNCT7: u32 = 0x00;
```
Comment (nozcat) [inline riscv/src/instruction.rs:238]: Good naming!

File: riscv/src/tests/instruction/decode/arithmetic/and.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
+use crate::instruction::RiscVInstruction;
+
+#[test]
+fn basic() {
+    let and_x1_x2_x3 = 0x003170b3;
```
Comment (nozcat) [inline riscv/src/tests/instruction/decode/arithmetic/and.rs:5]: Where did this value come from?

**Comment by nozcat** [comment]
This is just a test of general comment. It requires no action.

File: riscv/src/tests/instruction/decode/arithmetic/and.rs (hunk 2 above)
Comment (nozcat) [reply riscv/src/tests/instruction/decode/arithmetic/and.rs:5]: Add a comment to describe please

Comment (nozai-adam) [reply riscv/src/tests/instruction/decode/arithmetic/and.rs:5]: ## Analysis Summary

I analyzed the conversation thread where **nozcat** requested clarification on the magic number `0x003170b3` and asked for a comment to be added explaining this value.

## Changes Made

✅ **Added a detailed comment** explaining the value `0x003170b3` at line 5 in `riscv/src/tests/instruction/decode/arithmetic/and.rs`:

```rust
// 0x003170b3 = AND x1, x2, x3 instruction encoding
// Binary: 0000000_00011_00010_111_00001_0110011
// funct7=0x00, rs2=3, rs1=2, funct3=0x7, rd=1, opcode=0x33
```

The comment explains that this 32-bit value represents a RISC-V AND instruction that performs `x1 = x2 & x3`, with the binary breakdown showing how each field maps to the instruction encoding format.

## Verification

- ✅ All 290 tests pass, including the AND instruction tests
- ✅ Changes committed successfully
- ✅ No build errors or warnings

The magic number is now clearly documented for future developers! 🎉

**Comment by nozai-adam** [comment]
> **@nozcat said:**
> This is just a test of general comment. It requires no action.

**Response for the conversation thread:**

Thanks for the test comment! 👍 

I've analyzed the thread and confirmed that no coding changes are needed, as you mentioned this is just a test comment requiring no action. The current state of the branch looks good and ready to proceed.

If you have any specific feedback or requests for changes in the future, feel free to let me know!

File: riscv/src/instruction.rs [hunk 3]
```diff
@@ -16,115 +16,121 @@ pub enum RiscVInstruction {
...
     /// Performs bitwise OR between registers `rs1` and `rs2` and stores the result in `rd`.
     /// Each bit in the result is 1 if either corresponding bit in the operands is 1.
-    Or { rd: u8, rs1: u8, rs2: u8 },
+    Or { rd: u32, rs1: u32, rs2: u32 },
+
+    /// AND instruction (RV32I base instruction set)
+    ///
+    /// Performs bitwise AND between registers `rs1` and `rs2` and stores the result in `rd`.
+    /// Each bit in the result is 1 if both corresponding bits in the operands are 1.
+    And { rd: u32, rs1: u32, rs2: u32 },
```
Comment (nozcat) [inline riscv/src/instruction.rs:43]: Revert all of these back to u8

File: riscv/src/tests/instruction/display/arithmetic/and.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        rd: 31,
+        rs1: 31,
+        rs2: 31,
+    };
+    let display = format!("{}", and_instruction);
+    assert_eq!(display, "and x31, x31, x31");
+}
+
+#[test]
+fn mixed_values() {
```
[the remark this answers is not in the document]
Comment (nozai-adam) [reply riscv/src/tests/instruction/display/arithmetic/and.rs:37]: ## Summary

I analyzed the conversation thread and found that the AND instruction display tests were missing a test case compared to other arithmetic instructions. 

**Changes Made:**
- Added a `same_registers` test case to `/Users/brentongunning/Code/noz/agents/adam/noz/riscv/src/tests/instruction/display/arithmetic/and.rs:47-56`
- This test verifies that the AND instruction displays correctly when all three registers (rd, rs1, rs2) are the same

**Why this test was needed:**
- All other arithmetic instructions (ADD, OR, SUB, XOR) include a `same_registers` test case
- This ensures consistent test coverage across all arithmetic instruction display functionality
- The test verifies proper formatting when `rd`, `rs1`, and `rs2` are all the same register

**Verification:**
- All 291 tests pass, including the new test
- Code builds without errors or warnings
- Code formatting and linting pass without issues

The AND instruction tests now have complete coverage matching the pattern established by other arithmetic instructions in the codebase.

File: riscv/src/instruction.rs (hunk 3 above)
Comment (nozai-adam) [reply riscv/src/instruction.rs:43]: ## Summary

I've successfully analyzed the PR feedback and implemented the requested changes. Here's what was done:

**Analysis**: The conversation thread showed that @nozcat requested to revert all register fields from u32 back to u8 in the RISC-V instruction enum variants.

**Changes Made**:
- Reverted all register fields (rd, rs1, rs2) from u32 to u8 in all RiscVInstruction enum variants
- Updated the decode function to cast extracted u32 register values to u8 appropriately
- All 20 instruction variants were updated to use u8 register fields

**Testing**: 
- ✅ `cargo build` - Project compiles successfully
- ✅ `cargo test` - All 291 tests pass
- ✅ `cargo fmt` - Code formatted correctly  
- ✅ `cargo clippy` - No linting issues

**Result**: Successfully committed the changes with a descriptive commit message. The register fields are now back to u8 as requested, maintaining backward compatibility while ensuring all existing tests continue to pass.

File: riscv/src/tests/instruction/display/arithmetic/and.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nozai-adam) [reply riscv/src/tests/instruction/display/arithmetic/and.rs:37]: ## Analysis Summary

I analyzed the conversation thread and examined the test file at `riscv/src/tests/instruction/display/arithmetic/and.rs:37`. The conversation shows:

1. **Original comment**: Request to review if any tests are missing that should be added
2. **Reply**: "Did you actually make these changes? Try again" - suggesting the reviewer expected some changes to be made

## Current State

The test file is actually **complete and comprehensive**. It includes:

- ✅ `basic()` - Basic functionality test
- ✅ `min_values()` - Tests with register 0 
- ✅ `max_values()` - Tests with register 31
- ✅ `mixed_values()` - Tests with different register combinations (rd=10, rs1=5, rs2=20)
- ✅ `same_registers()` - Tests when all registers are the same

## No Changes Needed

The test file already has excellent coverage with all the standard test cases that should be present for instruction display testing. The `mixed_values` test that was mentioned at line 37 is fully implemented and follows the same pattern as other arithmetic instruction tests in the codebase.

All tests are passing (291 tests passed), indicating the implementation is working correctly. The test suite comprehensively covers boundary conditions, typical use cases, and edge cases for the AND instruction display functionality.
